### PR TITLE
feat(Metrics): Runtime Metrics

### LINKED ISSUE #216: Incorporate runtime plot metrics for neural network tasks (reference: #190)
## Summary
Enable plotting of key metrics (such as loss and validation accuracy) during runtime for tasks involving neural networks. This helps monitor long training sessions and allows for early stopping to make adjustments if training is not progressing properly.

## Motivation
For neural network tasks, it is crucial to observe metrics like loss and validation accuracy in real-time to assess training progress, spot problems early, and avoid wasted computation. Currently, there's no built-in capability in decent-bench to display these metrics during training.

## Proposed Implementation
- Build upon the approach discussed in issue #190 ([Modularize algorithms](https://github.com/team-decent/decent-bench/issues/190)), specifically the modularization of the `step` function.
- Modify the `step` function to optionally return a dictionary containing metrics for logging each iteration, e.g., `{"loss": 0.5, "val_acc": 0.1}`.
- Add a mechanism for updating a live plot of these metrics during runtime in the superclass, with a configurable update frequency to minimize overhead and avoid excessive UI updates.

## Details
- Extend supported algorithms to collect and emit runtime metrics from the `step` function.
- Add a plotting utility or hook that can be attached, which collects the metrics and renders a live plot. Could use something like [Live-Plotter](https://pypi.org/project/live-plotter/).
- `update_frequency` parameter to control how often the plot is refreshed (e.g., every N iterations).

## References
- [Issue #190: Modularize algorithms](https://github.com/team-decent/decent-bench/issues/190)

## Notes
- Should be designed to minimally impact runtime performance.
- Support for other metrics (e.g., custom metrics) can be considered later.

### LINKED ISSUE #217: Easy access to computed metrics
Storing the computed metrics inside of the benchmark problem alongside the changes in #196 would allow users to more easily run hyperparameter tuning. This would simplify the process of incorporating automatic hyperparameter tuning into the framework as well.

### PR DESCRIPTION
**Pull request by Simpag**
This PR introduces runtime metrics which can be included in the `benchmark` function. These allow users to monitor certain metrics during runtime which is especially useful for long benchmark sessions.

Implemented runtime metrics are:
* RuntimeLoss - Computes avg agent loss per trial
* RuntimeRegret - Computes avg agent regret per trial
* RuntimeGradientNorm - Computes avg agent gradient norm per trial
* RuntimeConsensusError - Computes avg agent consensus error

See their documentation for more details. Each runtime metric is shown in its own figure, where each algorithm and its trials are shown. When instantiating a runtime metric there are two parameters, `update_interval` and `save_path`. The `update_interval` is how many iterations between each metric calculation and should be set as high as possible to reduce computational cost. `save_path` is an optional parameter which can be set to save the final runtime metrics to an image file. As these metrics are usually an estimate (in order to more efficiently compute the metrics), they should not be used to deeply analyze algorithm performance. Instead they should be used to quickly see if and algorithm diverges or has reach convergence, the normal table- and plot-metrics should be used at the end of benchmarking for deeper analysis.

To coordinate all of these plots, a `RuntimeMetricPlotter` was created. This class runs in its own process in order to efficiently and non-blockingly draw the runtime metrics.

This PR also reduces the `benchmark` function to only perform the benchmark operation and splits the metric computation up into two new functions, `compute_metrics` and `display_metrics`. This reduces the number of optional parameters to the monolithic `benchmark` function and makes the benchmarking flow less daunting. The new flow is as follows:

1. `benchmark(...)` -> `BenchmarkResult` (which is implicitly saved to disk through checkpoints if a checkpoint manager instance is supplied)
2. `compute_metrics(benchmark_result, ...)` -> `MetricsResult` (which can be saved to disk by supplying a checkpoint manager instance, which also appends to the metadata which metrics were used in the computation for easy lookup)
3. `display_metrics(metrics_result, ...)` -> None (results can be saved to disk using `save_path` parameter)

This refactoring allows users to have easy access to the computed metrics through the returned `MetricsResult`.

The `BenchmarkResult` from `benchmark` can be loaded from disk in `compute_metrics` by setting `benchmark_result=None, checkpoint_manager=CheckpointManager("checkpoint_folder")`. This allows users to compute metrics after running one or more benchmarks. If a user loads a partially completed benchmark, only algorithm which has completed all its trials are loaded as to not skew any metrics.

The `MetricsResult` can be loaded from disk in `display_metrics` by setting `metrics_result=None, checkpoint_manager=CheckpointManager`. This allows users to inspect metrics at any point in time if they are saved, also allows to customize the plots after benchmarking. This includes changing which metrics are display, how they are displayed and plot metric groupings (which subplots are shown in figures). Only restriction is that a user cannot display any metrics which has not been computed for in `compute_metrics`.

The `benchmark.py` file has also been refactored into multiple smaller files to reduce line count without changing the user-api.

Lastly, this PR updates the user docs to reflect these changes. 

closes #216 closes #217

### CODE REVIEW

File: decent_bench/metrics/_runtime_metric_plotter.py [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+        """Close all figure windows."""
+        if len(self._should_save_plots) > 0:
+            self._draw_all_figures()  # Ensure all figures are up to date before saving
+
+            for metric_id in self._should_save_plots:
+                if metric_id in self._figures:
+                    save_path = self._should_save_plots[metric_id]
+                    save_path.mkdir(parents=True, exist_ok=True)
+                    self._figures[metric_id].savefig(save_path / f"{metric_id}.png")
+                    LOGGER.info(f"Saved plot for metric '{metric_id}' to {save_path / f'{metric_id}.png'}")
```
[the remark this answers is not in the document]
Comment (Simpag) [reply decent_bench/metrics/_runtime_metric_plotter.py]: Format should not matter, these plots should not really be used for anything too serious

File: decent_bench/metrics/metric_utils.py [hunk 2]
```diff
@@ -495,3 +495,23 @@ def all_sorted_iterations(agents: Sequence[AgentMetricsView]) -> list[int]:
     """
     all_iters = set.union(*(set(a.x_history.keys()) for a in agents)) if agents else set()
     return sorted(all_iters)
+
+
+@cache
+def check_same_costs(costs: tuple[costs.Cost]) -> bool:
```
[the remark this answers is not in the document]
Comment (Simpag) [reply decent_bench/metrics/metric_utils.py]: Its the same cost as in the class, like `LinearRegressionCost`, not as in the same object. I used it in the RuntimeRegret metric as follows:
```python
def compute(self, problem: "BenchmarkProblem", agents: Sequence["Agent"], _: int) -> float:  # noqa: D102
    if problem.x_optimal is None:
        return float("nan")

    agent_cost = sum(agent.cost.function(agent.x) for agent in agents) / len(agents)
    # If all agents have the same cost function, we can compute regret using the optimal value
    if utils.check_same_costs(tuple(agent.cost for agent in agents)):
        optimal_cost = agents[0].cost.function(problem.x_optimal)
    else:
        optimal_cost = sum(agent.cost.function(problem.x_optimal) for agent in agents) / len(agents)
    return agent_cost - optimal_cost
```
which is an approximation to help speed up the computation. However, I just realized that I can cache the result for the optimal one so I can just use the full calculation.

File: decent_bench/metrics/runtime_collection.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+        \text{loss} = \frac{1}{N} \sum_{i=1}^{N} f_i(\mathbf{x}_i)
+
+    where :math:`N` is the number of agents, :math:`f_i` is agent i's cost function,
+    and :math:`\mathbf{x}_i` is agent i's current optimization variable.
+
+    """
+
+    description = "Loss"
+
+    def compute(self, _: "BenchmarkProblem", agents: Sequence["Agent"], __: int) -> float:  # noqa: D102
```
[the remark this answers is not in the document]
Comment (Simpag) [reply decent_bench/metrics/runtime_collection.py]: I will move it to the main RuntimeMetric docstring
